compiler/optimizations/preNormalizeOptimizations.cpp

```diff
@@ -102,36 +117,256 @@ void doPreNormalizeArrayOptimizations() {
-    //
-    //  (call "may be local access" arrSymbol, idxSym0, ... ,idxSymN,
-    //                              paramControlFlag, paramStaticallyDetermined)
-    //
-    // we need to check the second argument from last to determine whether we
-    // are confirming this to be a local access or not
-    if (SymExpr *controlSE = toSymExpr(call->get(call->argList.length-1))) {
-      if (controlSE->symbol() == gTrue) {
-        return confirmAccess(call);
-      }
-      else {
-        return revertAccess(call);
+    ret = confirmed ? confirmAccess(call) : revertAccess(call);
+  }
+  else {
+    // maybe we have automatic aggregation but no automatic local access? In
+    // that case we may have generated one of this primitive, for aggregation
+    // detection, now we just remove it silently
+    ret = revertAccess(call);
+  }
+
+  INT_ASSERT(ret);
+
+  if (fAutoAggregation) {
+    AggregationCandidateInfo *aggCandidate = preNormalizeAggCandidate[call];
+    if (aggCandidate != NULL) {
+
+      if (fReportAutoAggregation) {
+        std::stringstream message;
+        message << "Aggregation candidate ";
+        message << "has ";
+        message << (confirmed ? "confirmed" : "reverted");
+        message << " local child ";
+        message << getForallCloneTypeStr(aggCandidate->forall);
+        LOG_AA(0, message.str().c_str(), aggCandidate->candidate);
       }
+
+      aggCandidate->logicalChildAnalyzed(call, confirmed);
     }
-    else {
-      INT_FATAL("Misconfigured PRIM_MAYBE_LOCAL_THIS");
+  }
+  return ret;
+}
+
+// called during LICM to restructure a conditional aggregation to direct
+// aggregation
+void transformConditionalAggregation(CondStmt *cond) {
+  // move the aggregation call before the conditional (at this point in
+  // compilation it must be inlined)
+  for_alist(expr, cond->elseStmt->body) {
+    cond->insertBefore(expr->remove());
+  }
+  
+  // remove the defpoint of the aggregation marker
+  SymExpr *condExpr = toSymExpr(cond->condExpr);
+  INT_ASSERT(condExpr);
+
+  if (fReportAutoAggregation) {
+    std::stringstream message;
+    message << "Replaced assignment with aggregation";
+    if (condExpr->symbol()->hasFlag(FLAG_AGG_IN_STATIC_AND_DYNAMIC_CLONE)) {
+      message << " [static and dynamic ALA clone] ";
     }
+    else if (condExpr->symbol()->hasFlag(FLAG_AGG_IN_STATIC_ONLY_CLONE)) {
+      message << " [static only ALA clone] ";
+    }
+    LOG_AA(0, message.str().c_str(), condExpr);
   }
-  return NULL;
+
+  condExpr->symbol()->defPoint->remove();
+
+  // remove the conditional
+  cond->remove();
 }
 
+// called after unordered forall optimization to remove aggregation code (by
+// choosing non-aggregated version)
+void cleanupRemainingAggCondStmts() {
+  forv_Vec(CondStmt, condStmt, gCondStmts) {
+    if (condStmt->inTree()) {
+      if (SymExpr *condSymExpr = toSymExpr(condStmt->condExpr)) {
+        //std::cout << "100\n";
+        if (condSymExpr->symbol()->hasFlag(FLAG_AGG_MARKER)) {
```

Comment:
It seems to me you could use a global variable like gTypeDefaultToken for this. All the symexprs could refer to the one global variable. You wouldn't need a flag or to worry about removing the Def for the global variable. Anyway the flag is also acceptable.

Reply:
Yeah, you have mentioned that before (but you didn't know about this context). I thought about it and this felt like something too specific, and maybe the approach you are proposing is more suitable for something that is widely used in the AST. I was on the fence, and as you sound OK with this, I am thinking of keeping things as-is, as I have already run some testing on this.

Reply:
I am keeping the implementation as-is.

With this implementation, we are adding `FLAG_AGG_IN_STATIC_ONLY_CLONE` or `FLAG_AGG_IN_STATIC_AND_DYNAMIC_CLONE` to the aggMarker symbol to make more sensible output for `--report-auto-aggregation`. If we were to have a single global symbol we won't be able to flag different aggregation conditionals.

There would have been ways to resolve that, of course (`PRIM_OPTIMIZATION_INFO` being one)